language-design.md

```diff
@@ -17,7 +17,17 @@ For each quote, describe what it was about the quote that led you pick it.
 
+I definitely disagreed with Pavlus' article, especially the claim quoted above. From my experience on the OCM project, I know that it can be hard to do studies on languages. Even though OCM was better from the standpoint of someone who was actually acquainted with parallel programming, it was hard to use novices to show this. User studies over the years have been problematic and it's very hard to teach people parallel programming over the course of half an hour or so. There are also a lot of problems with the idea of using novices to judge a language when they make up a very small portion of the userbase. Not all languages are designed to be easy to learn or to be very user friendly. In fact, I think it would be counterintuitive for most programming languages to have ease of use for novices as a primary concern. User-friendly languages will have to be more verbose than regular languages and will make doing just basic tasks a chore. In the example given in the article, a for loop in Quorum will take twice as many lines as C and about double to triple the number of characters. While languages like Quorum or Scratch may be more accessible to novices, they would likely hinder the productivity of most programmers in the long run. It's also worth noting that the paper only used a sample size of 18 people whose average age was 21 [[Stefik et al., 2011, p 5](http://ecs.victoria.ac.nz/foswiki/pub/Events/PLATEAU/2011Program/plateau2011-stefik.pdf)]. Even if the results were statistically significant, 18 people is a pretty small sample size. Their sample also probably don't represent many of the people learning to code, especially since many coding novices are significantly younger than 21.
+
+"I stand on this claim: I should not design a small language, and I should not design a large one. I need to design a language that can grow." [Steele, 1998, p 5]
+
+While the _Growing a Language_ article was interesting, I think the exact opposite advice applies when creating DSLs. In a general purpose programming language, growth is definitely important. You can't think of everything people will want to do with your language, and, even if you could, it would be overwhelming to include all that functionality. However, things are very different with a DSL. Since DSLs limit programmers to a specific domain, it's much easier to predict the functionality programmers will need to solve problems in that domain and to bake most of this functionality into the language from the beginning. This is highly desirable becuase people using DSLs want to work with abstractions that are natural, accurate, and relevant to them, but they don't want to (or can't) implement these abstrations themselves. While it may be useful to start off with a relatively small DSL, it will be important to have core features available from the start as programmers may not be able to add the features themselves and the DSL will likely be too limited to add the functionality from within its domain.
```

Comment:
I think this is true in general. I (personally) feel that the line between DSL's and general purpose PL's gets fudged from time to time, and in these cases, I wonder how much "growability" of a language matters. What do you think?
I think some DSL's lend themselves better to being grown, like Unix for terminal/command prompt, than others, like Context Free.